content/blog/2025-05-13-global-value-numbering/index.md

```diff
@@ -0,0 +1,81 @@
+
+### Value partitioning
+
+Instead of hashing expressions to values like local value numbering, value partitioning works by directly computing congruence classes of expressions, where two expressions are congruent if they have the same opcode all their arguments are congruent with each other. To perform value partitioning, we first put a program into SSA to ensure that each value has a unique variable associated with it.  We assume that all operations of a type are in the same congruence class, then repeatedly partition congruence classes where this cannot be true until we obtain a maximum fixed point.
+
+We implemented this algorithm for value partitioning, which was given in the paper:  
+```
+Initial partition: all values computed by the same opcode are in the same congruence classes
+
+worklist = classes in initial partitio
+while worklist is not empty:
+	select a class c from worklist
+	for each possible arg position p:
+		touched = ∅
+		for each value v:
+			if arg p of v is in c, add v to touched
+			for each class s where some but not all members are touched:
+				n = s & touched
+				s = s - n
+				if s in worklist:
+					add n to worklist
+				else:
+					add smaller of n and s to worklist
+  ```
+After this, we pick a representative for each congruence class, then replace every operation of that type with the representative.
+
+### Redundancy Elimination
+After standardizing our program to use values instead of expressions, we still need to convert this into a performance improvement. To do this, we use an available expressions dataflow analysis to calculate which values are available at each point in the program. Fortunately, the properties of the renaming algorithm make it very easy to define the analysis for calculating available expressions.
+
+- The initial input is the empty set.
+- The transfer function takes the union of a block's input and every expression in the block. If an expression already exists, it's redundant and can be removed.
+- The merge function takes the intersection of all the outputs of a block's predecessors.
+
+I also tried implementing partial redundancy elimination, which moves computations that are redundant along some execution paths back through the control graph to turn them fully redundant and optimize them away. This can be accomplished by performing global analyses to determine where computations can be safely moved and where moving them would save time, but I didn't have time to fully wrap my head around this and fix the bugs.
+
+  
+
+#### Implementation Notes
+
+Getting GVN right was very finicky and required reading the text very carefully. My biggest struggles in the end were first understanding the processing algorithm, then figuring out and debugging all the edge cases that arose from not reading the paper carefully enough.
```

Comment:
Can you expand on this a little bit? What are one or two examples of edge cases that you didn't anticipate? How about one debugging "war story" where you describe a problem that was hard to track down?